src/main/java/com/squareup/javapoet/AnnotationSpec.java

```diff
@@ -99,6 +107,60 @@ public static Builder builder(Class<?> type) {
     return builder(ClassName.get(type));
   }
 
+  public static AnnotationSpec get(AnnotationMirror annotation) {
+    AnnotationSpec.Builder builder = AnnotationSpec.builder(
+        ClassName.get((TypeElement) annotation.getAnnotationType().asElement()));
+    AnnotationSpecBuildingVisitor visitor = new AnnotationSpecBuildingVisitor(builder);
+    for (Map.Entry<? extends ExecutableElement, ? extends AnnotationValue> entry
+        : annotation.getElementValues().entrySet()) {
+      entry.getValue().accept(visitor, entry);
+    }
+    return builder.build();
+  }
+
+  public static class AnnotationSpecBuildingVisitor
+          extends SimpleAnnotationValueVisitor7<Builder,
+          Map.Entry<? extends ExecutableElement, ? extends AnnotationValue>> {
+
+    private final Builder builder;
+
+    public AnnotationSpecBuildingVisitor(Builder builder) {
+      super(builder);
+      this.builder = builder;
+    }
+
+    @Override
+    protected Builder defaultAction(Object o, Map.Entry<? extends ExecutableElement,
+                                                        ? extends AnnotationValue> entry) {
+      String name = entry.getKey().getSimpleName().toString();
+      return builder.addMember(name, CodeBlock.builder().add(entry.getValue().toString()).build());
+    }
+
+    @Override
+    public Builder visitAnnotation(AnnotationMirror a,
+                                   Map.Entry<? extends ExecutableElement,
+                                             ? extends AnnotationValue> entry) {
+      String name = entry.getKey().getSimpleName().toString();
+      return builder.addMember(name, "$L", get(a));
+    }
+
+    @Override
+    public Builder visitEnumConstant(VariableElement c,
+                                     Map.Entry<? extends ExecutableElement,
+                                               ? extends AnnotationValue> entry) {
+      String name = entry.getKey().getSimpleName().toString();
+      return builder.addMember(name, "$T.$L", c.asType(), c.getSimpleName());
+    }
+
+    @Override
```

Comment:
nit: we try to pack our signature lines onto fewer lines, not more. This one would be:

```
    @Override public Builder visitType(TypeMirror t,
        Map.Entry<? extends ExecutableElement, ? extends AnnotationValue> entry) {
```